### PR TITLE
[RLlib] New ConnectorV2 API #04: Changes to Learner/LearnerGroup API to allow updating from Episodes.

### PR DESCRIPTION
**Pull request by sven1977**
This PR introduces some major changes to the Learner and LearnerGroup APIs' `update()` methods.
This is in preparation for the rollout of the new ConnectorV2 API that will be run from within EnvRunners (env-to-module and module-to-env connector pipelines) as well as Learners (learner connector pipelines).

Changes:
* Disseminate the existing `update()` methods between `update_from_batch` vs `update_from_episodes()`.
* Using `update_from_episodes()` will allow the final train batch to be computed/compiled by the Learner (vs EnvRunner or `Algorithm.training_step()`), further offloading things to where they belong and reducing the mental load on the users (e.g. they should not be concerned with gathering data in the EnvRunners that is only later needed for training).
* Soft-deprecate (warning) the `Learner/LearnerGroup.update()` methods.
* An additional `self._postprocess_episodes()` method is called by Learner during the `update_from_...()` and can be overridden by specific algorithms. By default, this is simply a noop. This will allow us to pave the way for the Learner ConnectorV2 pipeline to perform possible preprocessing of the train data (batch or episodes).

## Why are these changes needed?

Deeper purpose:
This will enable PPO - once ConnectorV2 are used by EnvRunner and Learner - to fully separate value function predictions from the sampling phase. Instead, values predictions AND bootstrap value predictions will be performed solely on the Learner side (in distributed fashion, on GPU where available). PPO RLModule's `forward_exploration/inference` methods will then no longer be required to "think about" what the Learner might need and solely focus on action computations alone. Other algos will equally benefit from this separation of concerns.

As algos will be able to determine what special data they need from the sampled episodes (e.g. PG-style algos always need vf predictions), users will also be able to write custom (env-runner AND learner) connectors that transform the raw episode data into a data dict that's acceptable by their custom RLModule. For example, a user might provide an RLModule (independent of the algo used) that always requires the previous rewards. The user will then write a env-to-module as well as a learner connector that extracts this data from the episodes and builds it into the resulting action-computing- or train-batch.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/learner.py [hunk 1]
```diff
@@ -1228,13 +1228,16 @@ def additional_update_for_module(
 
     def update(
         self,
-        batch: MultiAgentBatch,
         *,
-        minibatch_size: Optional[int] = None,
-        num_iters: int = 1,
+        batch: Optional[MultiAgentBatch] = None,
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1232]: Happy to discuss the alternative to provide two different (mutually exclusive?) methods that the user/algo can decide to call: `update_from_batch` (for algos that do NOT require episode processing, such as DQN) or `update_from_episodes` (for algos that require a view on the sampled episodes for e.g. vf-bootstrapping, vtrace, etc..).

File: rllib/core/learner/learner.py [hunk 2]
```diff
@@ -1254,23 +1257,30 @@ def update(
...
             case `reduce_fn` is None and we have more than one minibatch pass.
         """
         self._check_is_built()
 
-        missing_module_ids = set(batch.policy_batches.keys()) - set(self.module.keys())
-        if len(missing_module_ids) > 0:
-            raise ValueError(
-                "Batch contains module ids that are not in the learner: "
-                f"{missing_module_ids}"
+        if batch is not None:
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1267]: In the alternative design (two update methods), we could avoid these rather ugly if-blocks then.

File: rllib/core/learner/learner.py [hunk 3]
```diff
@@ -1309,12 +1319,12 @@ def update(
                 metrics_per_module=defaultdict(dict, **metrics_per_module),
             )
             self._check_result(result)
-            # TODO (sven): Figure out whether `compile_metrics` should be forced
+            # TODO (sven): Figure out whether `compile_results` should be forced
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1322]: typo

File: rllib/core/learner/learner.py [hunk 4]
```diff
@@ -1309,12 +1319,12 @@ def update(
...
             #  to return all numpy/python data, then we can skip this conversion
             #  step here.
             results.append(convert_to_numpy(result))
 
-        batch = self._set_slicing_by_batch_id(batch, value=False)
+        self._set_slicing_by_batch_id(batch, value=False)
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1327]: `batch` never used.

File: rllib/core/learner/learner.py [hunk 5]
```diff
@@ -1330,6 +1340,34 @@ def update(
         # dict.
         return reduce_fn(results)
 
+    @OverrideToImplementCustomLogic
+    def _preprocess_train_data(self, *, batch, episodes) -> Tuple[Any, Any]:
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1344]: Not sure this should be private or public?

File: rllib/core/learner/learner_group.py [hunk 6]
```diff
@@ -146,25 +146,38 @@ def is_local(self) -> bool:
 
     def update(
         self,
-        batch: MultiAgentBatch,
         *,
-        minibatch_size: Optional[int] = None,
-        num_iters: int = 1,
+        batch: Optional[MultiAgentBatch] = None,
```
Comment (sven1977) [inline rllib/core/learner/learner_group.py:150]: same discussion as above.

File: rllib/core/learner/learner_group.py [hunk 7]
```diff
@@ -175,60 +188,98 @@ def update(
...
+                assert episodes is not None
+                results = self._get_results(
+                    self._worker_manager.foreach_actor(
+                        [
+                            partial(_learner_update, episodes_shard=episodes_shard)
+                            for episodes_shard in ShardEpisodesIterator(episodes, len(self._workers))
+                        ]
+                    )
+                )
+            # TODO (sven): Implement the case in which both batch and episodes might
```
Comment (sven1977) [inline rllib/core/learner/learner_group.py:245]: If we have mutually exclusive methods: `update_from_batch` and `update_from_episodes`, this case (both batch AND episodes provided by user) would not exist anyways.

File: rllib/core/learner/learner_group.py [hunk 8]
```diff
@@ -678,16 +761,14 @@ def load_module_state(
             # so we should not load any modules in the MARLModule checkpoint that are
             # also in the RLModule checkpoints.
             if modules_to_load:
-                if any(
-                    module_id in modules_to_load
-                    for module_id in rl_module_ckpt_dirs.keys()
-                ):
-                    raise ValueError(
-                        f"module_id {module_id} was specified in both "
```
Comment (sven1977) [inline rllib/core/learner/learner_group.py:686]: Bug: This error message will NOT have the correct missing `module_id`.

File: rllib/core/learner/learner.py [hunk 9]
```diff
@@ -1201,6 +1223,39 @@ def update(
         # dict.
         return reduce_fn(results)
 
+    @OverrideToImplementCustomLogic
```
Comment (kouroshHakha) [inline rllib/core/learner/learner.py:1226]: If there is any neural network inference, does it happen here or in the connector?

File: rllib/core/learner/learner.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/core/learner/learner.py:1106]: I like it if the two methods are separated. I don't think there would be a case where a specific algorithm's learner would have both methods implemented. ie DQN will only implemented `update_from_batch`, and PPO would only implement `update_from_episodes`. This is much much cleaner than mixing both into one function. The user will have to deal with less cognitive load if they are separated.

File: rllib/utils/minibatch_utils.py [hunk 10]
```diff
@@ -153,3 +155,69 @@ def __iter__(self):
             # TODO (Avnish): int(batch_size) ? How should we shard MA batches really?
             new_batch = MultiAgentBatch(batch_to_send, int(batch_size))
             yield new_batch
+
+
+@DeveloperAPI
+class ShardEpisodesIterator:
```
Comment (kouroshHakha) [inline rllib/utils/minibatch_utils.py:161]: can we have a unittest for this?

**Review by kouroshHakha — CHANGES_REQUESTED** [review]

File: rllib/core/learner/learner.py (hunk 9 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:1226]: Good question! The answer is: sometimes both.

For example: If you have some preprocessing needs for your training data (no matter whether episodes or batches), then you might want to do some preprocessing on this data (e.g. clip rewards, extend episodes by one artificial timestep for v-trace or GAE) and then perform a pre-forward pass through your network (e.g. to get the value estimates). For that pre-forward pass, you'll need to call your connector first to make sure this batch has all custom-required data formats (e.g. LSTM zero-padding). Only after all these preprocessing steps, you will be able to continue with the regular `forward_train` + loss + ... procedure.

File: rllib/core/learner/learner.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:1106]: I separated them in the LearnerGroup and Learner APIs:
- update_from_batch(async=False|True)
- update_from_episodes(async=False|True)

Comment (sven1977) [reply rllib/core/learner/learner.py]: Also, I think it's nicer to have the `async_update` bool option as an extra arg (instead of separate method) for better consistency and less code bloat.

**Review by kouroshHakha — APPROVED** [review]
